Rich text editor

sltsheryl opened the pull request:
- Used RichTextEditor from `'@mantine/rte'`;
- Saved the data from the rich text editor as HTML string instead of json becuz it's easier 
- (https://v2.mantine.dev/others/rte/ > Configure toolbar, can decide which tools we want to include)
- added styles.ts which is a hackish way cuz chakra resets the css so the data from dangerouslysetinnerhtml wasn't formatted properly 
- dompurify used to prevent xss attacks

<img width="1088" alt="Screenshot 2023-09-17 at 1 45 20 AM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/96589109/641ea334-d136-451b-875f-bae23c7ebe29">

# dark mode
<img width="1033" alt="Screenshot 2023-09-17 at 1 45 40 AM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/96589109/7790a12e-4594-4c57-93ef-bc1e20a83198">

# light mode
<img width="1041" alt="Screenshot 2023-09-17 at 1 45 57 AM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/96589109/fb04c0e9-ab40-4f5e-a948-2878f2f36d13">

sltsheryl committed:
feat: add rich text editor

sltsheryl committed:
style: change rich text editor fields

sltsheryl committed:
fix: fix styling

sltsheryl committed:
fix: sanitise html data with dompurify

sltsheryl committed:
fix: switch dark, light mode

sltsheryl committed:
fix: fix styling issues

sltsheryl committed:
Merge branch 'master' of github.com:CS3219-AY2324S1/ay2324s1-course-assessment-g17 into rich-text

sltsheryl committed:
refactor: modularise richtexteditor

sltsheryl committed:
feat: add blockquote

sltsheryl committed:
Merge commit '47f5503b40487c33824374393e71d913bec87be3' into rich-text

* commit '47f5503b40487c33824374393e71d913bec87be3':
  fix: import env properly
  [frontend] fix code editor language list
  [frontend] style: remove comment
  [frontend] fix: intellisense popover cut off
  [frontend] fix: code editor restrict file types
  [frontend] fix: remove console.logs
  [frontend] add code editor import confirmation
  [frontend] feat: code editor open file
  [frontend]: feat code editor download file
  [frontend] feat: toggle editor theme
  [frontend] feat: code editor allow code copy
  [frontend] feat: code editor toggle font size
  [frontend] feat: code editor lang selector
  [frontend] question indiv: add split pane
  [frontend] feat: set up basic code editor

sltsheryl committed:
fix: fix merge conflicts

sltsheryl committed:
fix: fix conflicts

sltsheryl committed:
style: fix quoteblock style

sltsheryl commented:
Could someone help with the formatting, currently if we use the rich text editor, it would be aligned to the center? But by default, it should be left. Thanks :)

<img width="409" alt="Screenshot 2023-09-18 at 10 42 34 AM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/96589109/dc5d1666-97bd-46d4-bb93-3da505ccac5c">

sltsheryl committed:
style: quoteblock

carriezhengjr commented:
This is what I see on the rich text editor branch, for one of the questions. This appears super nice:
<img width="707" alt="Screenshot 2023-09-20 at 22 33 39" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/97394017/a71c4d6d-b6f4-4f48-8048-a1ef69c381c3">

However, this is what I see on the rich text editor branch, for some of the other questions.
Taking "hello test" for example:
<img width="590" alt="Screenshot 2023-09-20 at 22 34 00" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/97394017/4e1f8392-5683-4e0a-ba45-397a783fe662">

For the same "hello test" question, this is what I see on other branches without the rich text editor:
<img width="729" alt="Screenshot 2023-09-20 at 23 08 24" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/97394017/3e74aeda-a5d9-496a-922c-dfe947f57466">

I'm not sure why, for some questions, the description section becomes misaligned on the rich text editor branch 🤔

sltsheryl committed:
fix: fix view question left align

sltsheryl commented:
@carriezhengjr 
fixed. the first one looks ok cuz the description is long, but yeah i changed it alr 
<img width="472" alt="Screenshot 2023-09-20 at 11 23 56 PM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/96589109/e7ab5c10-e980-44c0-a6ca-90e93192d356">

chloeelim commented on frontend/src/pages/questions/ViewQuestion.tsx:
```diff
@@ -46,31 +47,39 @@ const ViewQuestion: React.FC = () => {
+              <Heading as="h1" size="xl" textAlign="center">
+                {question.title}
               </Heading>
-              <Text whiteSpace="pre-line">{question.questionDescription}</Text>
```
Hm, should this `whiteSpace="pre-line"` be added back? Seems to be running into issues rendering the "\n" characters right now without this property 🤔

sltsheryl replied:
yeap i removed by accident

sltsheryl replied:
thanks :)

chloeelim commented on frontend/src/pages/questions/ViewQuestion.tsx:
```diff
@@ -46,31 +47,39 @@ const ViewQuestion: React.FC = () => {
+                <Heading as="h2" size="md">
+                  Description
+                </Heading>
+                <Text
+                  dangerouslySetInnerHTML={{
+                    __html: DOMPurify.sanitize(question.questionDescription),
+                  }}
+                />
```
wrt to the comment above

```suggestion
                <Text
                  whiteSpace="pre-line"
                  dangerouslySetInnerHTML={{
                    __html: DOMPurify.sanitize(question.questionDescription),
                  }}
                />
```

| `whiteSpace="pre-line"` |  without `whiteSpace="pre-line"`  |
|-|-|
| <img width="1437" alt="Screenshot 2023-09-20 at 11 43 53 PM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/79991214/ef7c5b76-b41e-44b8-b168-31d56629e47f"> | <img width="1413" alt="Screenshot 2023-09-20 at 11 44 08 PM" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/79991214/f1214940-58ba-4f63-bcbb-db002d140ca0"> |

chloeelim replied:
(oh this might be fixed in the prev commit)

chloeelim commented on backend/user-service/src/models/prisma/migrations/20230917193155_init2/migration.sql:
Not very familiar with prisma, just wondering whats the difference between this file and `models/prisma/migrations/20230917145045_init/migration.sql`? 😅

sltsheryl replied:
hmm im not sure if its correct to commit them both but what happened was i forgot to add unique constraint on the users table for username and email when i first set it up. so i added that, then created a migration, applied the migration and pushed the changes

chloeelim reviewed (commented):
Looks really cool and nice 🔥 !! Just have some questions :)

sltsheryl committed:
Update frontend/src/pages/questions/ViewQuestion.tsx

Co-authored-by: chloe lim <79991214+chloeelim@users.noreply.github.com>

sltsheryl committed:
remove align button

sltsheryl committed:
fix: fix blockquote

sltsheryl committed:
Merge branch 'master' into rich-text

* master:
  refactor: change edit button to indiv qn page, edit style of buttons, fix edit api url
  docs: include .env.docker to .gitignore
  feat: add docker-compose
  [frontend] feat: add 404 page
  fix: correct question icon file path
  changed button colours to gray, except arrow which is blue
  [frontend] feat: add edit button

# Conflicts:
#	backend/user-service/package-lock.json
#	backend/user-service/package.json
#	frontend/src/pages/questions/ViewQuestion.tsx

sltsheryl committed:
fix: resolve merge conflicts

chloeelim reviewed (approved):
lgtm! sorry for the delay!!
